3482 add `run` API for common training, evaluation and inference

Nic-Ma opened the pull request:
Task step 7 of https://github.com/Project-MONAI/MONAI/issues/3482 .

### Description
This PR added the `run` API to execute most common cases of supervised training, evaluation or inference from the config directly.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented:
This PR depends on #3822 .

Thanks.

Nic-Ma commented:
Hi @ericspod @wyli , Here I use Use the `distutils.util.strtobool()` to convert `string` to `bool`, maybe you guys have other better options?

Thanks in advance.

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @ericspod @wyli @rijobro ,

Due to the Blossom CI issue, we can't merge any PRs recently, this PR depends on #3822 but it's almost ready for review. Could you please help review this draft PR first?

Thanks in advance.

Nic-Ma commented:
/black

ericspod commented on monai/bundle/utils.py:
```diff
@@ -0,0 +1,141 @@
+        if filepath.lower().endswith((".yml", ".yaml")):
+            if "Loader" not in kwargs:
+                kwargs["Loader"] = yaml.FullLoader
+            return yaml.load(f, **kwargs)
```
Should use `yaml.safe_load` if we can, we're not using YAML's ability to construct Python objects with `!!python/object` syntax so it's safer to disallow it unless that creates issues.

Nic-Ma replied:
Sure, updated in the latest commit.

Thanks.

ericspod commented on monai/bundle/utils.py:
```diff
@@ -0,0 +1,141 @@
+    return parser[paths[1][1:] if paths[1] != "" else ""]
+
+
+def parse_id_value(pair: str) -> Tuple[str, Any]:
```
Maybe add another function `parse_id_dict` to use where a dictionary is being created:

```python
def parse_id_dict(pairs: Optional[Sequence[str]]) -> Dict[str, Any]:
    if pairs is None:
        return {}
    return dict(map(parse_id_value, pairs))
```

Nic-Ma replied:
Good point, added in the latest commit.

Thanks.

ericspod commented on tests/test_bundle_run.py:
```diff
@@ -0,0 +1,136 @@
+                # test override with the whole overriding file
+                json.dump("Dataset", f)
+
+            ret = os.system(
```
`os.system` isn't a reliable function to use because it depends on creating a new shell and the environment that has isn't well controlled. This will likely fail in certain Anaconda environments for example. It will also be sensitive to how the command is parsed which will vary by platform. I would recommend `subprocess.check_call` with `sys.executable` as the path to the interpreter itself so that you don't have to worry about what `python` is aliasing.

Nic-Ma replied:
Good point, updated in the latest commit.

Thanks.

Nic-Ma replied:
Hi @ericspod ,

When I changed to `subprocess`, the windows CI test can't pass:
https://github.com/Project-MONAI/MONAI/runs/5394963470?check_suite_focus=true#step:10:579
Maybe we need another command for windows platform?
```
======================================================================
ERROR: test_shape_0 (tests.test_bundle_run.TestBundleRun)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\parameterized\parameterized.py", line 533, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "D:\a\MONAI\MONAI\tests\test_bundle_run.py", line 134, in test_shape
    ret = subprocess.check_call(
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'C:\hostedtoolcache\windows\Python\3.8.10\x64\python.exe -m monai.bundle.scripts run --meta_file=C:\Users\RUNNER~1\AppData\Local\Temp\tmpnlqqp0ze\meta.yaml --config_file=C:\Users\RUNNER~1\AppData\Local\Temp\tmpnlqqp0ze\config.json --override='{postprocessing#<args>#transforms#2#<args>#output_postfix: seg, network: <file>C:\Users\RUNNER~1\AppData\Local\Temp\tmpnlqqp0ze\override1.JSON#move_net, dataset#<name>: <file>C:\Users\RUNNER~1\AppData\Local\Temp\tmpnlqqp0ze\override2.JSON}' --target=evaluator' returned non-zero exit status 1.
```

Thanks.

Nic-Ma replied:
I skipped the `subprocess` test for windows.

Thanks.

Nic-Ma commented on monai/bundle/run.py:
```diff
@@ -0,0 +1,55 @@
+
+    override = {}
+    if args.override is not None:
+        for pair in args.override:
```
Sure, updated in the latest commit.

Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @ericspod @wyli ,

Thanks for your review and suggestions.
I updated the PR for:
1. Removed the CLI entry, I think we don't need it so far, users can leverage many out-of-the-box CLI tool.
2. Added unit tests to leverage `google fire` CLI tool. It can't parse complicated args like "AA#BB#<args>" directly, so I still kept the `id_value_to_dict` utility function.
3. Added support to define the `run` args in a JSON or YAML file, can help simplify the command, refer to MLFlow: https://www.mlflow.org/docs/latest/projects.html#mlproject-file.
4. Changed file name to `scripts.py`, will also add the `schema verification` logic here in PR https://github.com/Project-MONAI/MONAI/pull/3865.

Do you think there is any missing features for the regular train / evaluate / inference `run`?

Thanks in advance.

wyli commented:
thanks, it looks good, one minor point for the command line, is it possible to have `python -m monai.bundle run` as the primary interface, and make "fire" an internal and optional API?

Nic-Ma commented on setup.cfg:
```diff
@@ -50,6 +50,8 @@ all =
     matplotlib
     tensorboardX
+    pyyaml
+    fire
```
Hi @wyli ,

I only use `fire` in the unit tests, is it necessary to add `fire` in this `setup.cfg` to make CI pass?

Thanks.

wyli replied:
this is for `pip install monai[fire]`, if this pip command is required then this .cfg is correct

Nic-Ma commented:
> thanks, it looks good, one minor point for the command line, is it possible to have `python -m monai.bundle run` as the primary interface, and make "fire" an internal and optional API?

Thanks for your review.
Sure, added support to use `monai.bundle.scripts` as entry with `fire` directly, and added the unit tests to cover it.
It still also support to call the `run()` method in other CLI tool.

Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @ericspod @wyli ,

I tried to define the abstract class for scripts, but seems we don't have enough usages so far, it's not clear and easy to extract the common features now. So I restored to put the essential features in functions directly, refer to MLFlow:
https://www.mlflow.org/docs/latest/_modules/mlflow/projects.html#run
Could you please help review the PR again?

Thanks in advance.

Nic-Ma commented:
/build

Nic-Ma commented:
/build

wyli commented on monai/bundle/utils.py:
```diff
@@ -0,0 +1,169 @@
+        # rejoin the rest, and remove blanks around value
+        value = ":".join(items[1:]).strip()
+
+    # try to convert the correct data type
```
This if tree could be replaced by ast.literal_eval()?

Nic-Ma replied:
Totally removed these logic in the latest commit.
Let's rely on `fire` or other CLI tools directly.

Thanks.

wyli commented on tests/test_bundle_run.py:
```diff
@@ -0,0 +1,181 @@
+
+from monai.transforms import LoadImage
+
+TEST_CASE_1 = [
```
Could you make this test case a JSON file in https://github.com/Project-MONAI/MONAI/tree/dev/tests/testing_data, perhaps adds a yaml version if needed...

Nic-Ma replied:
Added both JSON and YAML test files.

Thanks.

wyli commented on monai/bundle/scripts.py:
```diff
@@ -0,0 +1,88 @@
+    args_file: Optional[str] = None,
+):
+    """
+    Specify metadata file and config file to run a regular training or evaluation program.
```
I think we should move the examples to a readme file in the folder monai/bundle, with more examples and an overview of the syntax @, #

Nic-Ma replied:
Yeah, we definitely need to prepare a detailed document for all the features in `monai.bundle`.
Maybe start from @ericspod 's PR: https://github.com/Project-MONAI/MONAI/pull/3834

Thanks.

wyli commented on docs/source/bundle.rst:
```diff
@@ -32,3 +32,21 @@ Model Bundle
+-----------
+.. autofunction:: load_config_file
+
+.. autofunction:: load_config_file_content
```
Form a use's perspective I don't know the difference of load_config_file and load_config_file_content, parse_config_file. Could you revise all the function names in utilities if they are frequently used?

Nic-Ma replied:
Hi @wyli ,

Thanks for your suggestion, that's a good point.
I totally refactored the config reading logic in the latest commit.

Thanks.

wyli reviewed (commented):
Thanks, I put some minor comments, I can help update the docstrings in this PR after revising and let you review it.

Nic-Ma commented:
Hi @wyli @ericspod ,

Thanks for your review and feedback, I totally updated the PR based on comments and other reference projects.
Could you please help take a look again?
I agree we should write detailed user manual documentation later.

Thanks.

Nic-Ma commented:
/black

wyli commented:
`python -m monai.bundle run` works fine with the `__main__.py` file, there's a minor issue with `python -m monai.bundle run --help` described here https://github.com/google/python-fire/issues/76

ericspod commented on tests/test_bundle_run.py:
```diff
@@ -0,0 +1,82 @@
+            cmd = f"{sys.executable} -m fire monai.bundle.scripts run --meta_file {meta_file}"
+            cmd += f" --config_file {config_file} --override {{'evaluator#<args>#amp':False,{override}}}"
+            cmd += " --target evaluator"
+            ret = subprocess.check_call(cmd.split(" "))
```
I suspect using `split` is dividing the arguments incorrectly if there are spaces in any of the values you're passing in that should be kept as one item together. The `cmd` value should be built up as a list rather than as a string then broken up, so `cmd = [sys.executable, '"-m", "fire", "monai.bundle.scripts", "run", "--meta-file", meta_file, ...]`. This is one of those things that can sometimes vary by platform, paths in Windows have a tendency to have spaces in them more often than other OSs (ie. "Program Files") so splitting by space isn't reliable.

wyli commented on monai/bundle/scripts.py:
```diff
@@ -0,0 +1,95 @@
+        python -m monai.bundle run --override='{"net#<args>#ndims": 2}' ...
+
+        # Override some config values at runtime:
+        python -m monai.bundle run --"net#<args>#input_chns" 1 ...
```
Hi @Nic-Ma, I just added `**kwargs` to support this use case of `python -m monai.bundle run --"net#<args>#input_chns" 1`, do you have any concerns? also I'm not sure about `override: Optional[Union[Dict, str]]`, when is `str` input useful?

Nic-Ma replied:
`override` should not be `str`, that's outdated code and I forgot to update the typehints.

Thanks.

Nic-Ma replied:
Hi @wyli ,

I understand your proposal. But I feel it's a little bit confusing to have both `override` and `kwargs` args, maybe we can simply keep only one of them?
If only keeps `kwargs`, it can't support the case that the overriding keys happen to be "target", "config_file", "meta_file" or "args_file". And we may have more similar issues in other future scripts.
And may also consider supporting `kwargs` in the `args_file` file?
What do you think?

Thanks in advance.

wyli replied:
I reviewed the usage and found that providing a dict input at CLI is not that straightforward in terms of syntax '{"net#<args>#ndims": 2}'. The kwargs would make it simpler, I think we should keep it. Perhaps we remove 'override'? Please help look into this,(sometimes it's not override the existing options but adding new options?)

Nic-Ma replied:
OK, so I think we need to change `target` arg to other names like `runner_name`, `target` is slightly easier to conflict with some keys in the config content.

Thanks.

Nic-Ma commented on monai/bundle/config_parser.py:
```diff
@@ -170,38 +170,27 @@ def set(self, config: Any, id: str = ""):
-            return self._do_resolve(data)
+            path, ids = ConfigReader.split_path_id(config[len(self.macro) :])
+            parser = ConfigParser(config=self.get() if not path else ConfigReader.load_config_file(path))
+            return self._do_resolve(config=deepcopy(parser[ids]))
```
Hi @wyli ,

I wrote `data=...` separately in the previous code because we don't need `deepcopy` if the config is from file.
But anyway, I think it's not a problem to copy a small config content to simplify the code.

Thanks.

Nic-Ma commented on monai/bundle/config_reader.py:
```diff
@@ -24,123 +25,118 @@
+        path_name = cls.path_match.findall(path)
+        if not path_name:
+            return "", src  # the src is a pure id
+        if len(path_name) < 1 and len(path_name[0]) < 1:
```
I don't think this `if` will hapen, maybe we can simplify the logic to?
```py
        if not path_name:
            return "", src  # the src is a pure id
        return path_name[0][0], ids
```

Thanks.

wyli replied:
It'll be here for some corner cases for 'src'

Nic-Ma replied:
Sure, updated in the latest commit.

Thanks.

Nic-Ma commented on monai/bundle/config_reader.py:
```diff
@@ -0,0 +1,157 @@
+            src: source string to split.
+
+        """
+        path, *ids = f"{src}".rsplit(ReferenceResolver.sep, 1)
```
Hi @wyli ,

I feel this logic may not be correct, the `src` can be something like:
`"~/data/nic#work/config.json#net#<args>#channels"`
Maybe we should use `cls.path_match.findall` directly on `src`?

Thanks.

wyli replied:
Sure, I feel we need a refactoring for this to unify the ref/macro/I'd syntax, could you work on this?

Nic-Ma replied:
I updated this function logic in the latest commit.

Thanks.

Nic-Ma commented:
/black

wyli reviewed (approved):
thanks @Nic-Ma, this looks nice now and I'll revise and merge it soon.

wyli commented:
/build

Nic-Ma commented:
Hi @ericspod @wyli ,

Thanks for your review and updates.
I am trying to unify the `_name_` and `_path_` key words to `_target_`, similar to Hydra:
https://github.com/facebookresearch/hydra/blob/main/examples/instantiate/object/conf/db/mysql.yaml#L1

Thanks.

Nic-Ma commented on monai/bundle/config_item.py:
```diff
@@ -308,7 +310,7 @@ class ConfigExpression(ConfigItem):
     """
 
-    prefix = "$"
+    prefix = EXPR_KEY
```
Hi @wyli ,

If already defined the key in `utils`, do we still need this `prefix` member variable?
Same in all the other places.

Thanks.

Nic-Ma commented on monai/bundle/scripts.py:
```diff
@@ -0,0 +1,112 @@
+        ("meta_file", "config_file", "target_id", "args_file", "override"),
+        (meta_file, config_file, target_id, args_file, override),
+    )
+    print("\n--- input summary of monai.bundle.scripts.run ---")
```
Hi @wyli ,

Thanks for adding more printing information.
I feel it's better to leverage `logging` instead of `print` because users may trigger these scripts in complicated environments like FL, AutoML, etc.
Could you please help change it similar to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_writer.py#L34?

Thanks.

Nic-Ma commented on monai/bundle/scripts.py:
```diff
@@ -0,0 +1,112 @@
+        print(f"> {name}: {pprint.pformat(val)}")
+    print("---\n\n")
+
+    _args = _update_default_args(args=args_file, **override)
```
Hi @wyli ,

To print the run information, I think it's better to print the updated `_args` here instead of the `(meta_file, config_file, target_id, args_file, override)`.
What do you think?

Thanks.

Nic-Ma reviewed (commented):
Hi @wyli ,

Thanks for your update, I put some minor comments inline, others look good to me.

Nic-Ma commented:
/black
/build

wyli commented:
/build

wyli commented:
/build
